scan/vulnerability.go

```diff
@@ -0,0 +1,223 @@
+			dheCount = dheCount + 1
+		}
+	}
+	ciphers := make([]uint16, 0, dheCount)
+	for cipherID, name := range tls.CipherSuites {
+		if strings.Contains(name.Name, "_DHE_") {
+			ciphers = append(ciphers, cipherID)
+		}
+	}
+	return ciphers
+}
+
+func logjamScan(addr, hostname string) (grade Grade, output Output, err error) {
+	// Test 1: See if server will do a DHE handshake with a DHE_EXPORT cipher
+	tcpConn, err := net.Dial(Network, hostname+":443")
+	if err != nil {
+		return
+	}
+
+	config := defaultTLSConfig(hostname)
+
+	// Append DHE_EXPORT ciphers
+	ciphers := make([]uint16, 0, 2)
+	ciphers = append(ciphers, 0X0011)
+	ciphers = append(ciphers, 0X0014)
+	config.CipherSuites = ciphers
+
+	config.CurvePreferences = allCurvesIDs()
+	conn := tls.Client(tcpConn, config)
+	serverCipher, _, _, _, _, _ := conn.SayHello()
+	conn.Close()
+
+	// If Server accepted a DHE EXPORT cipher, then it is vulnerable.
+	if serverCipher != 0 {
+		grade = Bad
+		err = errors.New("Error! Server accepts a DHE EXPORT cipher")
+		return
+	}
+
+	// Test 2: Do a DHE handshake and look at the public exponent used by the server
+	tcpConn, err = net.Dial(Network, hostname+":443")
+	if err != nil {
+		return
+	}
+
+	config = defaultTLSConfig(hostname)
+	config.CipherSuites = allDHECiphersIDs()
+	config.CurvePreferences = allCurvesIDs()
+	conn = tls.Client(tcpConn, config)
+	serverCipher, _, _, _, _, _ = conn.SayHello()
+	conn.Close()
+
+	// If Server uses a DHE cipher with a public exponent <= 1024 bits, then it's vulnerable
+	if serverCipher != 0 {
+		certData, err := certinfo.ParseCertificateDomain(hostname)
+		if err != nil {
+			grade = Bad
+			return grade, output, err
+		}
+
+		ca, err := helpers.ParseCertificatePEM([]byte(certData.RawPEM))
+		if err != nil {
+			grade = Bad
+			return grade, output, err
+		}
+
+		// TODO: this public key isn't from the DHE handshake
+		// Not sure how to get that key...
+		pub, _ := ca.PublicKey.(*rsa.PublicKey)
+		if pub.N.BitLen() < 1024 {
+			grade = Bad
+			err = errors.New("Unsupported public key length")
+			return grade, output, err
+		}
+	}
+
+	grade = Good
+	return
+}
+
+func weakdhScan(addr, hostname string) (grade Grade, output Output, err error) {
+	// DHE handshake
+	tcpConn, err := net.Dial(Network, hostname+":443")
+	if err != nil {
+		return
+	}
+	config := defaultTLSConfig(hostname)
+	config.CipherSuites = allDHECiphersIDs()
+	config.CurvePreferences = allCurvesIDs()
+	conn := tls.Client(tcpConn, config)
+	conn.SayHello()
+	conn.Close()
+
+	err = errors.New("WIP")
+
+	// Look for Apache and other web servers' default primes
+	// Check DH param used in DHE against a set of standard ones shipped with web servers like Apache
+	// ???
+	return
+}
+
+// Helper to get all export grade ciphers
+func allExportCiphersIDs() []uint16 {
+	exportCount := 0
+	for _, name := range tls.CipherSuites {
+		if strings.Contains(name.Name, "_EXPORT_") {
+			exportCount = exportCount + 1
+		}
+	}
+	ciphers := make([]uint16, 0, exportCount)
```

Comment:
You can just define `var ciphers []uint16` and append onto it without counting out the capacity first.